freeipa/src/main/java/com/sequenceiq/freeipa/service/image/FreeIpaImageProvider.java

```diff
@@ -79,82 +77,79 @@ public List<ImageWrapper> getImages(ImageSettingsRequest imageSettings, String r
-    }
-
-    private List<Image> filterImages(List<Image> imageList, String os, String platform, String region) {
-        Predicate<Image> predicate = img -> img.getOs().equalsIgnoreCase(os) &&
-                img.getImageSetsByProvider().containsKey(platform) &&
-                filterRegion(region, platform, img);
-        Map<Boolean, List<Image>> partitionedImages =
-                Optional.ofNullable(imageList).
-                        orElse(Collections.emptyList()).
-                        stream()
-                        .collect(Collectors.partitioningBy(predicate));
-        if (hasFiltered(partitionedImages)) {
-            LOGGER.debug("Used filter for: | {} | Images filtered: {}",
-                    os,
-                    partitionedImages.get(false).stream().map(Image::toString).collect(Collectors.joining(", ")));
-            List<Image> images = partitionedImages.get(true);
-            return providerSpecificImageFilter.filterImages(platform, images);
+    private boolean filterRegion(FreeIpaImageFilterSettings imageFilterSettings, Image img) {
+        Set<String> regionSet = img.getImageSetsByProvider().get(imageFilterSettings.getPlatform()).keySet();
+        return CollectionUtils.containsAny(regionSet, imageFilterSettings.getRegion(), DEFAULT_REGION);
+    }
+
+    private List<Image> filterImages(List<Image> candidateImages, FreeIpaImageFilterSettings imageFilterSettings) {
+        LOGGER.debug("Filtering images with the following parameters: {}", imageFilterSettings);
+        String platform = imageFilterSettings.getPlatform();
+        List<Image> filteredImages = candidateImages.stream().filter(filterImagesPredicate(imageFilterSettings)).toList();
+        if (!filteredImages.isEmpty()) {
+            List<Image> notApplicableImages = new ArrayList<>(candidateImages);
+            notApplicableImages.removeAll(filteredImages);
+            LOGGER.debug("Used filter for: | {} | Images filtered: {}", imageFilterSettings.getOs(),
+                    notApplicableImages.stream().map(Image::toString).collect(Collectors.joining(", ")));
+            return providerSpecificImageFilter.filterImages(platform, filteredImages);
         } else {
-            LOGGER.warn("No FreeIPA image found with OS {}, falling back to the latest available one if such exists!", os);
-            return imageList;
+            LOGGER.warn("No FreeIPA image found with OS {}, falling back to the latest available one if such exists!", imageFilterSettings.getOs());
+            return candidateImages;
         }
     }
 
-    private boolean hasFiltered(Map<Boolean, List<Image>> partitioned) {
-        return !partitioned.get(true).isEmpty();
+    private Predicate<Image> filterImagesPredicate(FreeIpaImageFilterSettings imageFilterSettings) {
+        return image -> (imageFilterSettings.isAllowMajorOsUpgrade() || image.getOs().equalsIgnoreCase(imageFilterSettings.getOs())) &&
```

Comment:
this would allow rh8->centos7 or rh8->rh7 migration too. I'm not sure we should allow it